selfdrive/common/tests/test_util.cc

```diff
@@ -0,0 +1,64 @@
+
+#include <dirent.h>
+#include <sys/types.h>
+
+#include <algorithm>
+#include <climits>
+#include <random>
+#include <string>
+
+#define CATCH_CONFIG_MAIN
+#include "catch2/catch.hpp"
+#include "selfdrive/common/util.h"
+
+std::string random_bytes(int size) {
+  std::random_device rd;
+  std::independent_bits_engine<std::default_random_engine, CHAR_BIT, unsigned char> rbe(rd());
+  std::string bytes(size+1, '\0');
+  std::generate(bytes.begin(), bytes.end(), std::ref(rbe));
+  return bytes;
+}
+
+TEST_CASE("util::read_file") {
+  auto read_file = [](const std::string &fn) -> std::string {
+    std::ifstream f(fn);
+    std::stringstream buffer;
+    buffer << f.rdbuf();
+    return buffer.str();
+  };
+
+  SECTION("read /proc") {
+    DIR *d = opendir("/proc");
+    struct dirent *de = nullptr;
+    while ((de = readdir(d))) {
```

Comment:
Why not use `/proc/self/cmdline` so simplify the test? You can also simplify the test by just checking if it contains `test_util` instead of writing a 2nd implementation of `read_file`

Reply:
It is impossible to test that the open is successful, but tellg return -1 in this way

Reply:
It may be a file permission issue

Reply:
What situation are you trying to catch with this test? A file you don't have permission to? I don't like the randomness of loping over `/proc`. If you need something without read access try `/proc/kmsg`, or `/etc/shadow` if you also want to test with a regular file.

Reply:
There are indeed many problems, I need to learn how to write a good test case:D

Reply:
This looks so much cleaner :)